added non-blocking root communicator

gberg617 opened the pull request:
# Summary

This PR is a feature which adds a communicator for sending messages from any rank to the root rank non-collectively.  This can be useful in cases where an arbitrary rank throws an error that needs to be sent to the root rank to output to a file.

gberg617 commented:
Unit testing and documentation will be added to this PR in follow-up commits.

rhornung67 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
 
+const char* mpiNonBlockingReceiveMessages(MPI_Comm comm, const int tag)
+{
+  const int mpiTag = (tag == false) ? LJ_TAG : tag;
```
I think this should be changed to compare `tag` to an integer value not a bool. Also, the `const` qualifier is not needed on the tag argument because it is passed by value.

rhornung67 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
+  MPI_Status mpiStatus;
+
+  // Get size and source of MPI message
+  int mpiFlag = true;
```
Similar comment here.

gberg617 replied:
I ended up using integer values of 0 or 1 to indicate whether something is true or false.

rhornung67 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
                                 int destinationRank,
-                                const char* packedMessagesToBeSent)
+                                const char* packedMessagesToBeSent,
+                                const int tag)
```
```suggestion
                                 int tag)
```

rhornung67 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
+                                const char* packedMessagesToBeSent,
+                                const int tag)
 {
+  const int mpiTag = (tag == false) ? LJ_TAG : tag;
```
Similar comment to above.

rhornung67 commented on src/axom/lumberjack/MPIUtility.hpp:
```diff
@@ -30,6 +30,16 @@ namespace lumberjack
+ * \param [in] comm The MPI Communicator.
+ *****************************************************************************
+ */
+const char* mpiNonBlockingReceiveMessages(MPI_Comm comm, const int tag = 0);
```
Why is `tag` argument here?

gberg617 replied:
The MPI communication calls currently use the value associated with LJ_TAG by default (defined in MPIUtility.cpp).  The non-blocking receives used by the new communicator in this PR work better when we use another tag in order to not conflict with other communicators.  I added logic into the MPI utility functions to check whether the tag was overridden (i.e. non-zero).  In those cases, the sends/receives will use the tag value passed in.  Otherwise, we revert to the default LJ_TAG for MPI communication.  Setting this default in the function declarations prevents us from having to change all the existing calls to these methods by other communicators.

rhornung67 replied:
Ah. Got it. Thanks for the explanation.

bmhan12 replied:
Doxygen documentation for `tag` param should be added for `mpiNonBlockingReceiveMessages`, as done for `mpiNonBlockingSendMessages`

rhornung67 commented on src/axom/lumberjack/MPIUtility.hpp:
```diff
@@ -40,11 +50,14 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm);
                                 int destinationRank,
-                                const char* packedMessagesToBeSent);
+                                const char* packedMessagesToBeSent,
+                                const int tag = 0);
```
Similar comment about why `tag` arg is here.

rhornung67 commented on src/axom/lumberjack/NonBlockingRootCommunicator.cpp:
```diff
@@ -0,0 +1,91 @@
+        if (currPackedMessages == nullptr )
+        {
+          receive_messages = false;
+        } else {
```
Move opening curly brace to next line for formatting consistency.

gberg617 replied:
Fixed.

bmhan12 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
+    charArray = new char[messageSize + 1];
+    charArray[messageSize] = '\0';
+
+    // Receive packed Message
+    MPI_Recv(charArray,
+            messageSize,
+            MPI_CHAR,
+            mpiStatus.MPI_SOURCE,
+            mpiTag,
+            comm,
+            &mpiStatus);
```
As I understand the MPI API, this is actually a blocking `MPI_Recv` [call](https://www.mpich.org/static/docs/v3.4/www3/MPI_Recv.html)? So this `mpiNonBlockingReceiveMessages` function is currently blocking to receive messages.

gberg617 replied:
Yes, that's correct.  The non-blocking part is the call to MPI_Iprobe, but then the Recv is blocking.  My intent here is to be sure that the receive is fully finished before anything else is done, but to not block any further execution if there are no messages to be received (i.e. when mpiFlag is false).  I can change the function name to clarify the intent here.

gberg617 replied:
To clarify the above point, MPI_Iprobe is used instead of MPI_probe because the former will return with an mpiFlag value regardless of whether messages need to be received, whereas the latter is a blocking call that will only return if there is a message to be received.

gberg617 replied:
This does also make me think about renaming this communicator to something like "NonCollectiveCommunicator" rather than "NonBlockingCommunicator".  It's true that it calls these non-blocking functions, but I think the main feature is actually that we don't rely on collective calls to communicate messages to root.

rhornung67 replied:
I like that idea.

bmhan12 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +49,49 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
+
+  // Get size and source of MPI message
+  int mpiFlag = true;
+  MPI_Iprobe(MPI_ANY_SOURCE, tag, comm, &mpiFlag, &mpiStatus);
```
`MPI_Iprobe` is nonblocking here, so is there a chance the `mpiFlag` is not set to `true` when it is expected to be? Would it be better to have this be a blocking `MPI_Probe`? Basing this comment off this stackoverflow post: https://stackoverflow.com/questions/43823458/mpi-iprobe-vs-mpi-probe

Additionally, if using `MPI_Iprobe`, should `mpiFlag` default be set to `false`, so it can be set to `true` only by a successful function call?

gberg617 replied:
I think the mpiFlag will be set in either context to either true or false, but to your point, it is safer to initialize this as false.

gberg617 replied:
The stackoverflow example illustrates an interesting but slightly different approach than what I'm intending to do.  They are calling MPI_Iprobe in a while loop that does not exit until it returns a flag that is non-zero.  In my case, I am checking to see if any messages need to be received only once, and if there are no messages, the function exits by returning nullptr.  This intent in the stackoverflow example is to continuously monitor the status, whereas I'm only intending to periodically monitor the status whenever the code path enters into this function.  Both could be relevant to the problem I'm trying to solve with this communicator, where the root rank needs to receive information from other ranks that they are aborting. I had a preference toward the latter option (periodically monitoring the status whenever the root rank reaches a point where it enters this code path) because it seemed to me like the more efficient option, even if it comes at a cost of sometimes not receiving the status before the program aborts.  But I'm not really sure which option is best for this scenario.  I'd be curious to hear your thoughts.

bmhan12 replied:
>  I had a preference toward the latter option (periodically monitoring the status whenever the root rank reaches a point where it enters this code path) because it seemed to me like the more efficient option, even if it comes at a cost of sometimes not receiving the status before the program aborts.

I agree, I would expect the latter option to have less overhead, doing a single poll with `MPI_Iprobe` instead of spinning on `MPI_Iprobe` until status is updated in the former case. Nevertheless, I might not be considering something, so am also curious if others have ideas.

gberg617 replied:
I'm going to resolve this thread, because I believe we reached a consensus on this.

bmhan12 reviewed (approved):
LGTM!

white238 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,16 +48,50 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
   return charArray;
 }
 
+const char* mpiNonBlockingReceiveMessages(MPI_Comm comm, int tag)
```
Does it make sense to add `tag` to `mpiBlockingReceiveMessages`?

gberg617 replied:
This thread is no longer relevant, so I'll go ahead and resolve.

white238 commented on src/axom/lumberjack/MPIUtility.hpp:
```diff
@@ -40,11 +52,14 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm);
  * \param [in,out] packedMessagesToBeSent All of the Message classes to be sent
  *  packed together.
+ * \param [in] tag The MPI tag to use for communication.  When set to zero, 
+ *  MPI communication uses default LJ_Tag.
```
```suggestion
 *  MPI communication uses the default Lumberjack Tag.
```

white238 commented on src/axom/lumberjack/NonCollectiveRootCommunicator.cpp:
```diff
@@ -0,0 +1,99 @@
+  MPI_Comm_size(m_mpiComm, &m_mpiCommSize);
+  m_ranksLimit = ranksLimit;
+  m_mpiTag = mpiTag;
+  ++mpiTag;
```
Why is `mpiTag` being incremented here? This could use a comment.

gberg617 replied:
Each communicator that sends/receives messages non-collectively needs its own mpiTag in order to not interfere with other communicators.  I added this comment above.

gberg617 replied:
I would hope we don't have users who regularly initialize over 2 billion communicators :laughing: (although I've seen crazier things).  Anyways, I added the numeric_limits check in, and I reset the MPI tag to the initial value if it hits that limit.

gberg617 replied:
We switched to MPI_Comm_dup in the most recent commit, so the changes to the MPI tags are not relevant anymore.

white238 commented on src/axom/lumberjack/NonCollectiveRootCommunicator.hpp:
```diff
@@ -0,0 +1,143 @@
+}  // end namespace lumberjack
+}  // end namespace axom
+
+#endif
```
```suggestion
#endif

```

white238 commented on src/axom/lumberjack/tests/lumberjack_NonCollectiveRootCommunicator.hpp:
```diff
@@ -0,0 +1,106 @@
+          found = true;
+        }
+      }
+      if(!found)
+      {
+        std::cout << "Error: Message not received: " << currMessage << std::endl;
+      }
```
```suggestion
      SLIC_ERROR_IF(!found, "Message not received: " << currMessage);
```

white238 commented on src/axom/lumberjack/tests/lumberjack_NonCollectiveRootCommunicator.hpp:
```diff
@@ -0,0 +1,106 @@
+          found = true;
+        }
+      }
+      if(!found)
+      {
+        std::cout << "Error: Message not received: " << currMessage << std::endl;
+      }
+      EXPECT_EQ(found, true);
```
```suggestion
      EXPECT_EQ(found, true) << "Message not received: " << currMessage << std::endl;
```

white238 reviewed (changes requested):
The axom team needs to discuss the implications of this further before this is merged.

gberg617 commented:
@white238 @bmhan12 I talked with @gunney1 and we decided the best path forward is to duplicate the MPI communicator passed in the initialize() call, and have this duplicate owned by the Lumberjack communicator object.  With this change, we can avoid having to create MPI tags for each non-collective communicator object, and instead have each communicator object have its own MPI communicator using the same default MPI tag.  Please let me know if you have any further concerns.

kennyweiss commented on src/axom/lumberjack/MPIUtility.hpp:
```diff
@@ -30,6 +30,18 @@ namespace lumberjack
+ *  MPI communication uses default LJ_Tag. 
+ *****************************************************************************
+ */
+const char* mpiNonBlockingReceiveMessages(MPI_Comm comm);
```
Should we update the docs for this function (and `mpiBlockingReceiveMessages`) to indicate that the caller is responsible for deallocating the memory in the returned pointer?

gunney1 replied:
@gberg617 We should update docs as @kennyweiss suggested.

gberg617 replied:
Added a sentence about this to the function documentation,

kennyweiss reviewed (commented):
Thanks @gberg617 

Sounds like we'll need to discuss some details before merging this. I've added a few minor comments in the meantime.

gunney1 commented on src/axom/lumberjack/MPIUtility.hpp:
```diff
@@ -30,6 +30,18 @@ namespace lumberjack
+/*!
+ *****************************************************************************
+ * \brief Receives any Message sent to this rank, if there are any messages 
+ *  that are sent. Returns null if no messages are sent.
```
Since we use a non-blocking receive, there could be sent messages that haven't arrived yet.  I suggest replacing "are sent" with "have arrived."  It's pedantic but can avoid confusion when the unexpected happens.

gberg617 replied:
Fixed.

gberg617 commented:
Update on the failing tests: The multiple_communicators test I added sporadically fails on Azure.  This test is important to capture certain behavior when multiple communicators are sending/receiving messages.  Looking into this, it sometimes fails when run with other unit tests (and only on Azure), but consistently passes when it is run by itself.  I have heard that gtest+MPI is somewhat fragile in some cases, so I attempted to remove gtest from the tests I added by converting the ASSERT macros to my own version, and having each test become a function that is called within `main()` inside of `lumberjack_NonCollectiveRootCommunicator.cpp`.  This fixes the issue, and both tests in this file always pass.  Does the Axom team have any thoughts about potentially adding unit tests that do not use gtest?  If there is a strong preference for requiring gtest, does anyone have thoughts on other solutions that can prevent these sporadic failures?

bmhan12 commented:
> Update on the failing tests...

For reference, this is the error that is seen on Azure Pipelines and when ran locally in Docker:
```
102: Test command: /usr/bin/mpirun "-np" "7" "/home/axom/axom/build-clang@14.0.0-release/tests/lumberjack_mpi_tests" "--gtest_filter=lumberjack_NonCollectiveRootCommunicator*"
...
...
...
102: /home/axom/axom/src/axom/lumberjack/tests/lumberjack_NonCollectiveRootCommunicator.hpp:121: Failure
102: Expected equality of these values:
102:   receivedPackedMessages_c1.size()
102:     Which is: 0
102:   1
102: 
102: 
102: lumberjack_mpi_tests:8831 terminated with signal 11 at PC=55a67bf82d09 SP=7ffe05eea260.  Backtrace:
102: /home/axom/axom/build-clang@14.0.0-release/tests/lumberjack_mpi_tests(+0x8d09)[0x55a67bf82d09]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS_4TestEvEET0_PT_MS4_FS3_vEPKc+0x48)[0x7f8043294ef8]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing4Test3RunEv+0xcf)[0x7f80432790ef]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing8TestInfo3RunEv+0xf9)[0x7f804327a3d9]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing9TestSuite3RunEv+0x285)[0x7f804327aeb5]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing8internal12UnitTestImpl11RunAllTestsEv+0x41c)[0x7f804328b81c]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing8internal35HandleExceptionsInMethodIfSupportedINS0_12UnitTestImplEbEET0_PT_MS4_FS3_vEPKc+0x48)[0x7f8043295c18]
102: /home/axom/axom/build-clang@14.0.0-release/lib/libgtest.so.1.13.0(_ZN7testing8UnitTest3RunEv+0x59)[0x7f804328b3c9]
102: /home/axom/axom/build-clang@14.0.0-release/tests/lumberjack_mpi_tests(main+0x35)[0x55a67bf83075]
102: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f804015dd90]
102: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f804015de40]
102: /home/axom/axom/build-clang@14.0.0-release/tests/lumberjack_mpi_tests(+0x4f15)[0x55a67bf7ef15]
102: 
102: ===================================================================================
102: =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
102: =   PID 8832 RUNNING AT e17873b86e18
102: =   EXIT CODE: 9
102: =   CLEANING UP REMAINING PROCESSES
102: =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
102: ===================================================================================
102: YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
102: This typically refers to a problem with your application.
102: Please see the FAQ page for debugging suggestions
```

bmhan12 commented on src/axom/lumberjack/tests/lumberjack_NonCollectiveRootCommunicator.hpp:
```diff
@@ -0,0 +1,154 @@
+
+  if(commSize > 1 && c1.rank() == 0)
+  {
+    c1.push(nullptr, receivedPackedMessages_c1);
```
> Update on the failing tests: The multiple_communicators test I added sporadically fails

@gberg617 I was able to fix the sporadic failure locally in my Docker container by doing the following where the root rank expects to receive messages:
```suggestion
    while(receivedPackedMessages_c1.size() == 0)
    {
      c1.push(nullptr, receivedPackedMessages_c1);
    }

```
To explain why I think the test was failing the check (this doesn't explain why the application terminates, still don't understand that):

- A non-blocking send is sent by rank 1. We don't know for sure when this completes.
- The ranks synchronize at a `MPI_Barrier`. **Notably**, this _does not guarantee_ that the non-blocking send has completed at the exit of the barrier.
- Rank 0 attempts to receive the non-blocking send, the send has not completed, and size 0 is returned, failing the check.

The added while loop lets the root rank poll until it sees the single message.

I think the reason this error didn't pop up on LC systems is because the inter and intra connects are so much faster than on a standard laptop/CI node.

gberg617 commented on src/axom/lumberjack/NonCollectiveRootCommunicator.cpp:
```diff
@@ -0,0 +1,96 @@
+    {
+      currPackedMessages = mpiNonBlockingReceiveMessages(m_mpiComm);
+
+      if(isPackedMessagesEmpty(currPackedMessages))
```
I believe `isPackedMessagesEmpty()` will check to see if `currPackedMessages` is null.

gberg617 commented on src/axom/lumberjack/NonCollectiveRootCommunicator.hpp:
```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and
```
Fixed.

gberg617 commented on src/axom/lumberjack/NonCollectiveRootCommunicator.hpp:
```diff
@@ -0,0 +1,135 @@
+   *
+   * \param [in] comm The MPI Communicator
+   * \param [in] ranksLimit Limit on how many ranks are individually tracked per
+   *  Message.
```
Fixed.

gberg617 commented on src/axom/lumberjack/NonCollectiveRootCommunicator.cpp:
```diff
@@ -0,0 +1,96 @@
+  MPI_Comm_dup(comm, &m_mpiComm);
+  MPI_Comm_rank(m_mpiComm, &m_mpiCommRank);
+  MPI_Comm_size(m_mpiComm, &m_mpiCommSize);
+  m_ranksLimit = ranksLimit;
```
I put in a std::cerr to avoid including anything from SLIC in Lumberjack.

gberg617 commented:
All the major issues have been resolved.  Let me know if there are any other thoughts on this MR.

gunney1 commented on src/axom/lumberjack/MPIUtility.hpp:
```diff
@@ -33,7 +34,8 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm);
  *****************************************************************************
  * \brief Receives any Message sent to this rank, if there are any messages 
- *  that are sent. Returns null if no messages are sent.
+ *  that have arrived. Returns null if no messages are sent. Caller is 
```
Another place where "are sent" should be replaced with "have arrived".

gunney1 reviewed (approved):
Please first address the change in MPIUtility.hpp.

bmhan12 commented:
ReadTheDocs documentation update for this PR can be found here: https://axom.readthedocs.io/en/feature-bergel1-lumberjack_nonblocking_communicator/axom/lumberjack/docs/sphinx/communicator_class.html

white238 commented on RELEASE-NOTES.md:
```diff
@@ -43,6 +43,7 @@ to use Open Cascade's file I/O capabilities in support of Quest applications.
 - Adds intersection routines between `primal::Ray` objects and `primal::NURBSCurve`/`primal::NURBSPatch` objects.
 - Adds LineFileTagCombiner to Lumberjack to allow combining messages if line number, file, and tag are equal.
 - Adds some support for 2D shaping in `quest::IntersectionShaper`, using STL meshes with zero for z-coordinates or in-memory triangles as input.
+- Adds NonCollectiveRootCommunicator to Lumberjack to provide an MPI-based communicator for logging messages non-collectively.
```
```suggestion
- Adds `NonCollectiveRootCommunicator` to Lumberjack to provide an MPI-based communicator for logging messages non-collectively.
```

white238 commented on src/axom/lumberjack/MPIUtility.cpp:
```diff
@@ -49,6 +48,37 @@ const char* mpiBlockingReceiveMessages(MPI_Comm comm)
+    charArray[messageSize] = '\0';
+
+    // Receive packed Message
+    MPI_Recv(charArray,
```
`MPI_Recv` is the blocking version:

https://www.mpich.org/static/docs/v3.4/www3/MPI_Recv.html

Is there something im missing?

gberg617 replied:
Yes, exactly.  It's Iprobe and subsequent mpiFlag check that makes this non-blocking.  I can add a comment somewhere in this code to clarify that point.

white238 replied:
@kennyweiss @rhornung67 

I spoke to @gberg617 in the hall and my hangup is with the name of this function not the functionality. It does a blocking receive only if there is a message, otherwise it does nothing but the probe.

Should this be called something like `mpiOptionalReceiveMessages` or `mpiBlockingReceiveifMessagesExist`? Or maybe just add a parameter to the existing version?

gberg617 replied:
I like your second suggested name.  I can go ahead and change it to that if you'd like.

gberg617 replied:
Fixed! I can go ahead and mark this resolved if you'd like.

bmhan12 commented on src/axom/lumberjack/docs/sphinx/communicator_class.rst:
```diff
@@ -50,3 +50,18 @@ This Communicator has all nodes directly connecting to the root node which
+is rank 0.  The root node is the only node allowed to output messages.
+Each single push, the child nodes send their currently held messages
+to the root.  Unlike the RootCommunicator above, the messages are pushed by
+individual ranks rather than collectively by all ranks.  The root rank
+receives messages sent from other ranks with this communicator when it
+calls push.
```
It would be good to add a warning here to remind users that this communicator is non-collective, that it is up to them to verify all the sent messages have been received by the root rank.

gberg617 replied:
I added a warning to the end of this section clarifying that point.

gberg617 replied:
Marking as resolved.

gberg617 commented:
@white238 @bmhan12 @rhornung67 @kennyweiss @gunney1 I added an example in Lumberjack called `lumberjack_noncollective_messages_ex` that represents a typical use case where a non-collective root communicator is needed to log error messages.  I attempted to re-create the same set of conditions encountered in the host code of interest, but it is still somewhat simplified.  I try to emphasize what this example is supposed to illustrate with the detailed comments in the source code (see `nonCollectiveMessagesExample.cpp`).  Feel free to let me know if you have any thoughts/suggestions.
